guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+## Avoid `beforeEach` and `afterEach`
+
+When writing tests that need to be set up or torn down in a similar way, it may be tempting to extract the setup and teardown phases to `beforeEach` and `afterEach` hooks:
+
+🚫
+
+``` typescript
+describe('TokensController', () => {
+  let controller: TokensController;
+  let onNetworkDidChangeListener: (networkState: NetworkState) => void | undefined;
+
+  beforeEach(() => {
+    onNetworkDidChange = jest.fn();
+    controller = new TokensController({
+      chainId: '0x1',
+      onNetworkDidChange: (listener) => {
+        onNetworkDidChangeListener = listener;
+      },
+    });
+  });
+
+  afterEach(() => {
+    controller.destroy();
+  });
+
+  describe('addToken', () => {
+    it('registers the given token under the default network, assuming it has not changed yet', () => {
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x1': {
+          symbol: 'DAI',
+        },
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      onNetworkDidChangeListener!({ chainId: '0x42' });
+
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x42': {
+          symbol: 'DAI',
+        },
+      });
+    });
+  });
+});
+```
+
+However, this strategy ends up increasing the long-term maintenance cost of tests for a couple of reasons:
+
+- It makes debugging tests difficult later. Data that is critical to the purpose of a test — i.e., that is used in the execution or verification phase — may be hidden in a `beforeEach` hook, or it may be split across `beforeEach` and the test in question. This means that in order to trace data as it moves through the test, one must spend time piecing together the full picture.
+- It makes writing new tests difficult. Extracting the setup phase assumes that all tests will need to be set up in exactly the same way. But the very nature of code guarantees that this will not be the case. Inevitably, tests will involve scenarios that cause differences in behavior from the "happy path". Using `beforeEach`, in order to construct a new scenario, the test must extend what `beforeEach` is already doing, whether this is passing a different argument to the function being tested, mocking a function to return a different result, or something else. But this is not always as straightforward as it sounds. It is common to declare variables above the `beforeEach` which can be used later on in the file, but the utility of these variables is misleading, as by the time a test runs, these variables have already been used to set it up and thus cannot be reused or changed. Sometimes this creates enough of a problem that developers resort to copying the entirety of the setup code from `beforeEach` into the test, but this introduce another way that maintainers must now use to work with the tests.
```

Comment:
Nit: I get what you were going for here, but there's a lot going on here. I've tried to simplify it a bit. I found the examples later in this line to be too confusing as well - I think we'd need to give them more space if we want to keep them in.

```suggestion
- It makes writing tests for new scenarios difficult. The setup steps in `beforeEach` may not apply to future tests that cover different scenarios. In these cases, the author is forced to take on a complex refactor to remove the `beforeEach` steps that don't apply, or pursue complex workarounds.
```

Reply:
You're right. I like your new wording, that sounds a lot better 😅 Changed here: [054f0c4](https://github.com/MetaMask/contributor-docs/pull/38/commits/054f0c43fa2488e6f92a7a6192970a21a3169029)